containers/hawkular-openshift-security-filter/src/main/java/org/hawkular/openshift/auth/org/hawkular/openshift/namespace/NamespaceHandler.java

```diff
@@ -0,0 +1,130 @@
+
+/**
+ *
+ */
+public class NamespaceHandler implements HttpHandler {
+
+    private static final Logger log = Logger.getLogger(NamespaceHandler.class);
+
+    private final HttpHandler containerHandler;
+
+    //
+    private static final String HAWKULAR_TENANT_HEADER_NAME = "Hawkular-Tenant";
+    private static final String TOKEN_FILE_LOCATION = "/var/run/secrets/kubernetes.io/serviceaccount/token";
+
+    private static final String KUBERNETES_MASTER_URL_SYSPROP = "KUBERNETES_MASTER_URL";
+    private static final String KUBERNETES_MASTER_URL_DEFAULT = "https://kubernetes.default.svc.cluster.local";
+    static final String KUBERNETES_MASTER_URL =
+            System.getProperty(KUBERNETES_MASTER_URL_SYSPROP, KUBERNETES_MASTER_URL_DEFAULT);
+    //
+
+    private final String token;
+
+    private final NamespaceListener namespaceListener;
+
+    private final Map<String,String> projectOverwrites;
+
+    public NamespaceHandler(HttpHandler containerHandler) {
+        log.info("Creating NamespaceHandler");
+        this.containerHandler = containerHandler;
+
+        try {
+            this.token = getToken();
+            log.info("Received token from " + TOKEN_FILE_LOCATION);
+        } catch (Exception e){
+            log.error("Error retrieving token from file " + TOKEN_FILE_LOCATION);
+            throw new RuntimeException(e);
+        }
+
+        namespaceListener = new NamespaceListener(token);
+        NamespaceOverrideMapper nom = new NamespaceOverrideMapper(token);
+        log.info("Received namespace overrides: " + nom.overrides);
+
+        this.projectOverwrites = nom.overrides;
+    }
+
+    private String getToken() throws IOException {
+        return new String(Files.readAllBytes(Paths.get(TOKEN_FILE_LOCATION)));
+    }
+
+    @Override
+    public void handleRequest(HttpServerExchange serverExchange) throws Exception {
+        String originalTenantName = getTenantName(serverExchange);
+        if (originalTenantName == null || originalTenantName.startsWith("_")) {
+            containerHandler.handleRequest(serverExchange);
+        } else {
+
+            StringJoiner namespaceName = new StringJoiner(",");
+            String[] splitted = originalTenantName.split(",");
+
+            for (String namespace: originalTenantName.split(",")) {
+                String namespaceID = namespaceListener.getNamespaceId(namespace);
+
+                if (namespaceID != null) {
+                    String tenantName;
+
+                    if (projectOverwrites.containsKey(namespaceID)) {
+                        tenantName = projectOverwrites.get(namespaceID);
+                    } else {
+                        tenantName = namespace + ":" + namespaceID;
+                    }
+                    namespaceName.add(tenantName);
+
+
+                } else {
+                    log.error("Could not determine a namespace id for namespace " + namespaceID + ". Cannot process request. Returning an INTERNAL_SERVER_ERROR.");
```

Comment:
I would reduce logging level to info here. Can you use the other version of ```endExchange``` that allows you to set a message in the response. The message could be the same as the log message.